File: apps/android-unity/Assets/Game/Core/Random/SplitMix64.cs
using System;

namespace Baseball.Core.Random
{
    /// <summary>Bit-for-bit port of SimulationCore.SplitMix64.</summary>
    public struct SplitMix64
    {
        public SplitMix64(ulong seed)
        {
            State = seed;
        }

        public ulong State { get; private set; }

        public ulong Next()
        {
            unchecked
            {
                State += 0x9E3779B97F4A7C15UL;
                var value = State;
                value = (value ^ (value >> 30)) * 0xBF58476D1CE4E5B9UL;
                value = (value ^ (value >> 27)) * 0x94D049BB133111EBUL;
                return value ^ (value >> 31);
            }
        }

        public int NextInt(int upperBound)
        {
            if (upperBound <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(upperBound), "upperBound must be positive");
            }

            return (int)(Next() % (ulong)upperBound);
        }
    }
}

File: apps/android/game-core/src/main/kotlin/com/solkim/baseball/core/SplitMix64.kt
package com.solkim.baseball.core

/** Bit-for-bit port of the current C#/Swift SplitMix64 implementation. */
public class SplitMix64(public var state: ULong) {
    public fun next(): ULong {
        state += 0x9E3779B97F4A7C15UL
        var value = state
        value = (value xor (value shr 30)) * 0xBF58476D1CE4E5B9UL
        value = (value xor (value shr 27)) * 0x94D049BB133111EBUL
        return value xor (value shr 31)
    }

    public fun nextInt(upperBound: Int): Int {
        require(upperBound > 0) { "upperBound must be positive" }
        return (next() % upperBound.toULong()).toInt()
    }
}
